### PR TITLE
feat(ads): Beckhoff ADS plugin — connectivity (2/6)

### PR DESCRIPTION
**Pull request by RuneRoven**
Connect, route creation, reconnect, Close. Adds the go-ads→benthos slog bridge (`benthosLogHandler`): trace suppressed, severities mapped, verbosity governed solely by the pipeline log level (no global `SetDefaultLogger`). Port fix: host-IP auto-detect TCP dial follows `targetPort` (UDP discovery stays 48899).

---
Part of the ENG-4538 Stage-1 stack (re-cut of the old #350–#354 into the 6 ticket buckets). **Merge as one batch, bottom-up (1→6).** logLevel config removed (ADS verbosity now follows the pipeline log level); connect auto-detect dial follows targetPort. CI for the unit tests is deferred to the next stage.

### CODE REVIEW

**Comment by RuneRoven** [comment]
@coderabbitai review

File: beckhoff_ads_plugin/ads.go [hunk 1]
```diff
@@ -268,10 +268,67 @@ func init() {
...
+	// Ensure done channel is initialized (may be nil if constructed directly in tests)
+	if g.done == nil {
+		g.done = make(chan struct{})
+	}
+	g.Log.Infof("Creating new connection")
+	opts, err := g.buildSessionOptions(ctx)
+	if err != nil {
+		return err
+	}
+	if err := g.openSession(ctx, opts); err != nil {
```
Comment (led0nk) [inline beckhoff_ads_plugin/ads.go]: just style but lets keep the approach with first assigning and then checking on the error here as well

File: beckhoff_ads_plugin/ads.go [hunk 2]
```diff
@@ -268,10 +268,67 @@ func init() {
...
+		if g.Handler != nil {
+			_ = g.Handler.Close()
```
Comment (led0nk) [inline beckhoff_ads_plugin/ads.go]: lets make heare use of something like 
```golang
        defer func() {
                if err != nil {
                        g.closeHandler()
                }
        }()
```

where then `g.closeHandler()` has the guard included itself

File: beckhoff_ads_plugin/connect.go [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+	return opts, nil
+}
+
+// resolveRouteHostIP returns the local IP for the ADS route: the configured
+// hostIP, else the source IP of a TCP dial to the ADS port (same interface).
+func (g *AdsCommInput) resolveRouteHostIP(ctx context.Context) (string, error) {
+	if g.HostIP != "" {
+		return g.HostIP, nil
+	}
+	dialer := net.Dialer{Timeout: 3 * time.Second}
```
Comment (led0nk) [inline beckhoff_ads_plugin/connect.go]: for all those things i'd like to do a const-block somewhere - could also lilve in constants.go and then write a small (1 -line) coment why its 3 seconds

also for e.g. targetPort a few lines below the initial-samples and so on

File: beckhoff_ads_plugin/logging.go [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+	attrs  []slog.Attr
+}
+
+func (h *benthosLogHandler) Enabled(_ context.Context, level slog.Level) bool {
+	// Suppress trace (LevelTrace = -8); forward debug and above to benthos,
+	// which then applies its own configured level filter.
+	return level >= slog.LevelDebug
+}
+
+func (h *benthosLogHandler) Handle(_ context.Context, r slog.Record) error {
```
Comment (led0nk) [inline beckhoff_ads_plugin/logging.go:37]: as said, we need to make sure to not introduce another logger here and make use of the benthos-umh one. If a function e.g. returns an error wwe could make use of that and use the benthos-logger but don't inject / wrap or do sth else with the logging

**Review by led0nk — CHANGES_REQUESTED** [review]
i wrote a smole gist for this in general which would make the library on the later stage easier swappable. 

https://gist.github.com/led0nk/7e679c9ae206a56b3e4a37fc49d9672a

The reamde is generated but take a look more on the code, which was an idea from my perspective. The relevant part here is that we reconstruct some of the struct and make use of the interface so we can later on easily swap out things here. Then we also can redruce the points of imports for the library here. 

Btw some things might get better naming e.g. `goADSClient` could also simply be Client or sth like this

File: beckhoff_ads_plugin/logging.go (hunk 4 above)
Comment (RuneRoven) [reply beckhoff_ads_plugin/logging.go:37]: go-ads is slog-based and always logs: if you pass no logger, client.go:136 defaults to slog.Default(). So dropping the bridge doesn't remove logging — go-ads's connection/route/notification diagnostics then spew to slog.Default() → raw stderr, bypassing benthos's pipeline and level control.

bridge → benthos routes the go-ads slog to g.Log with the benthos log level.

This simply bridges all logging direct into benthos to get the same format and not the stderr output.

File: beckhoff_ads_plugin/ads.go (hunk 2 above)
Comment (RuneRoven) [reply beckhoff_ads_plugin/ads.go]: - closeHandler() (read.go) closes ASYNC — go func(){ _ = c.Close() }(). That's deliberate for the read hot-path (ReadBatchPull/ReadBatchNotification):
  go-ads Close() can block on protocol teardown, and a synchronous close there would stall the Benthos read loop. Async is right there.
  - Connect's failure cleanup closes SYNC — g.client.Close() directly. That's right for a setup path: on a failed connect Benthos will retry Connect, and
  we want the dead session fully torn down before returning, so the retry starts clean.
  
  (Connect→initSymbolIndex→loadSymbolTable→setupNotifications) into one finishConnect, so there's a single error path and one explicit close — already
  DRY, nothing to dedupe with a defer.
  - newGoADSClient failure needs no cleanup (client not yet assigned → still nil) — handled by only cleaning up after g.client = c.

File: beckhoff_ads_plugin/ads.go [hunk 5]
```diff
@@ -263,17 +257,82 @@ func init() {
...
 
-func (g *AdsCommInput) Connect(_ context.Context) error {
-	return service.ErrNotConnected
+// Connect establishes the ADS session, resolves symbol metadata, and (depending
+// on config) loads the full symbol table and registers notifications.
+func (g *AdsCommInput) Connect(ctx context.Context) error {
+	if g.client != nil {
+		return nil
+	}
+	if g.done == nil {
```
Comment (led0nk) [inline beckhoff_ads_plugin/ads.go]: check if we really need to do this here, since we created the channel ourseelves earlier on, otherwise we could then only do it here e.g.

File: beckhoff_ads_plugin/client_goads.go [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+
+// goADSClient is the go-ads adapter — the ONLY file (besides client.go)
+// importing the fork. All WithRoute/NewAMSAddress/type-mapping logic lives here.
+type goADSClient struct {
+	session *adsLib.Session
+}
+
+// isLikelyContainerIP flags private ranges commonly used by container/pod
+// networks, where route registration would advertise an unreachable IP.
+func isLikelyContainerIP(ip string) bool {
```
Comment (led0nk) [inline beckhoff_ads_plugin/client_goads.go]: let's challenge if we really need to have this here in the first place.

what if one comes up with an ip network starting with `10.0.0.0` here? 
we have it in the office and it would be recognized as potential container ip, then i'm questioning to have this in the first place. It makes sense to simply check the same way as in the configuration PR on the IP. So whether this is parsable to a IPv4 in the first place, but then i'd think of that it makes sense to reuse the one of the configuration PR

File: beckhoff_ads_plugin/ads.go [hunk 7]
```diff
@@ -263,17 +257,82 @@ func init() {
...
+		g.client = nil
+		return err
+	}
+	return nil
+}
+
+// finishConnect drives connect → symbol resolution → (optional) symbol table
+// load → (optional) notification setup; split out so Connect can clean up g.client on failure.
+func (g *AdsCommInput) finishConnect(ctx context.Context) error {
+	g.Log.Infof("Connecting to plc")
```
Comment (led0nk) [inline beckhoff_ads_plugin/ads.go]: maybe let's print out more information here, e.g.:
- targetIP + port
- hostAMS + port
could also be multiline if it get's too long

File: beckhoff_ads_plugin/ads.go [hunk 8]
```diff
@@ -263,17 +257,82 @@ func init() {
...
+		}
+	}
+	return nil
 }
 
 func (g *AdsCommInput) ReadBatch(_ context.Context) (service.MessageBatch, service.AckFunc, error) {
 	return nil, nil, service.ErrNotConnected
 }
 
 func (g *AdsCommInput) Close(_ context.Context) error {
```
Comment (led0nk) [inline beckhoff_ads_plugin/ads.go]: maybe for closing the connection we should as well state some information, such as (example):
`Closing connection to (ip-address):(port)`

`Connection successfully closed`

This seems a bit unusual but in later stages it sometimes helps on debugging

File: beckhoff_ads_plugin/client_goads.go [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+	// CGNAT range used by some Kubernetes CNIs: 100.64.0.0/10
+	if p[0] == 100 && p[1] >= 64 && p[1] <= 127 {
+		return true
+	}
+	return false
+}
+
+// resolveRouteHostIP returns the local IP for the ADS route: the configured
+// hostIP, else the source IP of a TCP dial to the ADS port (same interface).
+func resolveRouteHostIP(ctx context.Context, cfg SessionConfig, log *service.Logger) (string, error) {
```
Comment (led0nk) [inline beckhoff_ads_plugin/client_goads.go:37]: just in order to understand this correctly here:

- we do a tcp-port dial
- > if it fails we do a udp-port dial
- > other than that we report back the ip address and append this with `1.1` outside of this? 

Is this the automatic route creation then?

File: beckhoff_ads_plugin/connect.go [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+
+// loadSymbolTable downloads the full symbol and datatype table from the PLC;
+// required for struct/array symbols.
+func (g *AdsCommInput) loadSymbolTable(ctx context.Context) error {
+	g.Log.Infof("Loading symbol and datatype table from PLC (loadSymbols=true)")
+	if err := g.client.LoadSymbols(ctx); err != nil {
+		g.Log.Errorf("LoadSymbols failed: %v", err)
+		return err
+	}
+	g.Log.Infof("Symbol table loaded")
```
Comment (led0nk) [inline beckhoff_ads_plugin/connect.go]: Seems nitty but lets do the logging here consistent, maybe sth like:

Loading symbol and datatype table from PLC...

-> error case: `Loading symbol and datatype failed:...`

successcase: `Loading symbol and datatype succeeded.`

File: beckhoff_ads_plugin/connect.go [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+
+	// AddNotifications returns nil errors even when all symbols fail to
+	// resolve (e.g. PLC not yet ready); detect that here so Benthos retries Connect().
+	registered := 0
+	for _, r := range results {
+		switch {
+		case r.Registered:
+			registered++
+		case r.Skipped:
+			g.Log.Errorf("Notification symbol %q skipped (check symbol name)", r.SymbolName)
```
Comment (led0nk) [inline beckhoff_ads_plugin/connect.go]: if i understand correctly we error out if registered notifications are 0

then we will try to reconnect right?
If so i personally would demote this here to a Warning, as well as the default case here.

Then on a complete error we will in the if loop there return the error which than get logged outside of this function as an error and throws the reconnection error right?

File: beckhoff_ads_plugin/connect.go [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+	for len(needed) > 0 {
+		select {
+		case update := <-g.NotificationChan:
+			if update != nil {
+				g.pendingInitial = append(g.pendingInitial, update)
+				delete(needed, strings.ToLower(sampleName(update)))
+			}
+		case <-initialCtx.Done():
+			g.Log.Warnf("Timed out waiting for initial samples; %d symbols not yet received: %v", len(needed), needed)
+			return nil
```
Comment (led0nk) [inline beckhoff_ads_plugin/connect.go:136]: won't this happen from a bad connection or so? doesn't it then make sense to return the initialCtx.Err() here as well if so?

File: beckhoff_ads_plugin/logging.go (hunk 4 above)
Comment (led0nk) [reply beckhoff_ads_plugin/logging.go:37]: unresolving this just so that you see this message:

for the future what we want to have is:

- the library doesn't do the logging for us
- the library supports all methods and always returns specific errors
- based on the errors, we will do the log-handling within benthos-umh
- this will then make this here redundant and we can get rid of this

**Review by led0nk — CHANGES_REQUESTED** [review]

File: beckhoff_ads_plugin/client_goads.go (hunk 6 above)
Comment (RuneRoven) [reply beckhoff_ads_plugin/client_goads.go]: this is mainly if you run standalone benthos and put it in a container. Using it in the umh-core it will always be a container so it depends on how you think about it. I tried to find ways to check what the actual host ip was or find out if running inside a container to throw a warning based on that but i could not find any good solution. If running standalone benthos you can derive the hostAMS from the actual host ip but inside a container you will always need to specify the hostIP and derive the hostAMS from that. For umh-core we could change the hostIP input to required but then you will need to specify this when running standalone which i don't like. We can for sure remove it and enhance the documentation about this.

File: beckhoff_ads_plugin/ads.go (hunk 5 above)
Comment (RuneRoven) [reply beckhoff_ads_plugin/ads.go]: You are correct, i removed it.

File: beckhoff_ads_plugin/client_goads.go (hunk 6 above)
Comment (RuneRoven) [reply beckhoff_ads_plugin/client_goads.go]: I leave it for now but i remove the 10.x.x.x and made a range array with adrresses to check instead so it is easy to change. Could remove the entire thing if we think that is is the best solution

File: beckhoff_ads_plugin/client_goads.go (hunk 9 above)
Comment (RuneRoven) [reply beckhoff_ads_plugin/client_goads.go:37]: Yes, this is used in the route creation. I will run an additional test again on this to validate this. The route creation and response id/ip is not well documented so much of this is reverse engineered. It might work with container ip as the netid.

Comment (RuneRoven) [reply beckhoff_ads_plugin/client_goads.go:37]: I have done some more testing now and updated my internal documents about this. It seems to work from container ip in TC3 but not in TC2. So for now we will leave this and the isLikelyContainerIP. It looks like TC2 uses the hostIP as unique identifier and TC3 uses the netID. I will later deep dive in this when i implement the secure ADS.

File: beckhoff_ads_plugin/connect.go (hunk 11 above)
Comment (RuneRoven) [reply beckhoff_ads_plugin/connect.go]: Yes, if all fails to register we error out and reconnect. Changed this to a warning and i group the fails to and array and then loop the warnings.

File: beckhoff_ads_plugin/connect.go (hunk 12 above)
Comment (RuneRoven) [reply beckhoff_ads_plugin/connect.go:136]: a  registered symbol always delivers a first sample if everything is correct.
Did some more testing for this, and found an edge case with the  fixed 10s wait against a user-configurable `maxDelay`. With `maxDelay: 20s`: warning at 10.0s, first sample at 14.7s — the PLC may hold a notification for up to cycleTime + maxDelay, so the gate expired before delivery was due. Now derived from the slowest symbol (`cycleTime + maxDelay + initialSampleWait`);  same config, no timeout. Defaults go 10s -> 10.2s.

A bad connection will not get here unnoticed.  `client.Connect` or `AddNotifications` would have failed first, and one that dies during the wait is caught by `IsClosed()` -> `ErrNotConnected` on the next read. So lets keep the `nil`. Erroring would re-register every second forever  and l block the other symbols.

File: beckhoff_ads_plugin/client_goads.go (hunk 6 above)
Comment (led0nk) [reply beckhoff_ads_plugin/client_goads.go]: honestly i think removing this is the best way here, because one could have a very weird docker networking setup and we will never be able to predict this.

Then i think we should not at any circumstance "restrict" this and even check for this.

I'd personally stay with this configoption but not this check here

File: beckhoff_ads_plugin/client_goads.go (hunk 9 above)
Comment (led0nk) [reply beckhoff_ads_plugin/client_goads.go:37]: this IMO is more evidence on the targetIP/Port thing which gets parsed by a helper function in the first place from my perspective.

But other than that i think we could make things "more clear", which also gets back to the config.
So having something that basically states us here "tc2/tc3" then one would only need to hand over the address in this function here as well, because whenever this here is called you could have a switch running on which type of the protocol were using here.

Not a hard opinion, but reduces a bit complexity and makes the code more readable in regards of "ah this is the tc2 path"

File: beckhoff_ads_plugin/logging.go (hunk 4 above)
Comment (led0nk) [reply beckhoff_ads_plugin/logging.go:37]: btw, forgot about this - this is maybe stuff for the rollout-ticket

**Review by led0nk — CHANGES_REQUESTED** [review]
looks good to me, but would prefer to solve the things that are still open (the logging thing is just to take note on the ticket for future plans)
